### PR TITLE
Feature/gap 2152 v2 notifications service

### PR DESCRIPTION
**Pull request by john-tco**
[GAP-2152](https://technologyprogramme.atlassian.net/jira/software/projects/GAP/boards/216?selectedIssue=GAP-2152)

- Updates cron functions to fetch email address from user-service
- Removes the existing jwt based unsubscribe logic; adds a new unsubscribe reference table
- updates node version - (develop currently red due because it's now out of date - https://github.com/cabinetoffice/gap-find-backend/actions/runs/6611768238/job/17956323518)

related prs: https://github.com/cabinetoffice/gap-find-web/pull/45, https://github.com/cabinetoffice/gap-user-service/pull/135

[GAP-2152]: https://technologyprogramme.atlassian.net/browse/GAP-2152?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-16.14.0
\ No newline at end of file
+18.12.0
```
Comment (dominicwest) [inline .nvmrc:1]: I won't lie - updating the node version our app uses scares me deeply 🤣

Comment (john-tco) [reply .nvmrc:1]: Yeah this change literally breaks typeorm and needs reverted. `yarn set version berry` seems to depend on this version now as it flags the following error (on the develop branch too)

Usage Error: This tool requires a Node version compatible with >=18.12.0 (got 16.14.0). Upgrade Node, or set `YARN_IGNORE_NODE=1` in your environment.

Comment (dominicwest) [reply .nvmrc:1]: As discussed, lets revert the node version until we have e2e tests in place to have some sort of confidence. Will do some digging on the pipeline flow, but can go with that fix for now to get around it

File: src/migration/1697725143044-addUnsubscribeReferenceTable.ts
Comment (dominicwest) [inline src/migration/1697725143044-addUnsubscribeReferenceTable.ts:1]: Probably worth preparing all the migrations that we're going to need to apply to our envs on confluence

File: src/notifications/notifications.service.ts [hunk 2]
```diff
@@ -55,11 +51,11 @@ export class NotificationsService {
                 'GOV_NOTIFY_SAVED_SEARCH_NOTIFICATION_EMAIL_TEMPLATE_ID',
             );
         this.HOST = this.configService.get<string>('HOST');
-        this.FRONT_END_HOST = this.configService.get<string>('FRONT_END_HOST');
     }
 
     async processGrantUpdatedNotifications() {
         console.log('Running Process Grant Updated Notifications...');
+        console.log('Running Process Grant Updated Notifications.. OLD ONE.');
```
Comment (dominicwest) [inline src/notifications/notifications.service.ts:58]: Remove one of these logs / make it clearer plz

File: src/notifications/v2/notifications.helper.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        this.USER_SERVICE_URL =
+            this.configService.get<string>('USER_SERVICE_URL');
+        this.SUBSCRIPTIONS_PER_BATCH = this.configService.get<number>(
+            'SUBSCRIPTIONS_PER_BATCH',
+        );
+        this.LAMBDA_SECRET = this.configService.get<string>('LAMBDA_SECRET');
+    }
+
+    async getUserServiceEmailsBySubBatch(batchOfSubs: string[]) {
+        console.log(this.LAMBDA_SECRET);
```
Comment (dominicwest) [inline src/notifications/v2/notifications.helper.ts:41]: Lets perhaps not log this haha
```suggestion
 
```

**Review by dominicwest — COMMENTED** [review]
Obviously finding it very difficult to review this, would have ideally seen a PR per scheduled job or something like that to make life easier! Also hard to determine whats copied code and whats new.

I'm assuming we've ran all of these jobs locally and they're working fine? If not will set this up locally and attempt to do so

But from what I can see it looks good, was clearly a massive effort getting this working. Especially when its an entirely new framework you both had to learn. Well done lads

File: src/notifications/notifications.service.ts (hunk 2 above)
Comment (john-tco) [reply src/notifications/notifications.service.ts]: yeah sorry this was just a leftover debug log, removed

File: src/notifications/v2/notifications.savedSearch.service.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            console.log(`Number of emails sent: ${notifications.length}`);
+            await this.savedSearchNotificationService.deleteSentSavedSearchNotifications();
+            console.log(
+                `saved search notifications temp table has been cleared`,
+            );
+            console.log(
+                `Task took ${
+                    performance.now() - startTime
+                } milliseconds to run \r\n`,
+            );
```
Comment (dominicwest) [inline src/notifications/v2/notifications.savedSearch.service.ts:186]: Some of these logs will be really unclear when debugging.
1. Its logging inside of a loop so will be spammy/unclear
2. Doesn't reference the scheduled job its running under
3. `saved search notifications temp table has been cleared` doesn't feel significant enough to log unless of course theres an error

File: src/notifications/v2/notifications.service.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            [ScheduledJobType.SAVED_SEARCH_MATCHES]:
+                this.v2SavedSearchService.processSavedSearchMatches,
+            [ScheduledJobType.SAVED_SEARCH_MATCHES_NOTIFICATION]:
+                this.v2SavedSearchService
+                    .processSavedSearchMatchesNotifications,
+        };
+        const cronFn = CRON_JOB_MAP[type as keyof typeof CRON_JOB_MAP];
+        const cronJob = getCronJob(cronFn, timer);
+        this.schedularRegistry.addCronJob(`${type}_${index}`, cronJob);
+        cronJob.start();
```
Comment (dominicwest) [inline src/notifications/v2/notifications.service.ts:36]: At first this confused me, but now I understand it love this refactor 👌

File: src/notifications/v2/notifications.savedSearch.service.ts [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                `Number of saved saved searches to process: ${
+                    savedSearches ? savedSearches.length : 0
+                }`,
+            );
+            console.log(
+                `Number of saved search notifications created: ${numberOfSearchesWithMatches}`,
+            );
+            console.log(
+                `Task took ${endTime - startTime} milliseconds to run \r\n`,
+            );
```
Comment (dominicwest) [inline src/notifications/v2/notifications.savedSearch.service.ts:151]: Would ideally reference the scheduled job in each log, easy for them to get lost amongst other things happening concurrently in prod

File: src/notifications/v2/notifications.grant.service.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+                    batchesCount,
+                );
+                await this.sendGrantUpdatedEmailsToBatch(
+                    batch as Subscription[],
+                    grantId,
+                );
+            }
+        }
+
+        await this.contentfulService.updateEntries(grantIds, {
```
Comment (dominicwest) [inline src/notifications/v2/notifications.grant.service.ts:213]: This function makes a lot of logs, would get rid of it logging the whole contentful entry and make it a bit clearer what its doing. Again, reference the scheduled job its running under in the logs

**Review by dominicwest — CHANGES_REQUESTED** [review]
Just a comment in general - for all logs related to these jobs, reference the job name its running under. And some additional logs around numbers affected and things like that would be ideal. There's a couple excessive logs too that output a whole contentful blob which is not ideal. Appreciate a lot of this was copy and paste, but lets improve things so its easier to debug if anything goes wrong

File: src/notifications/v2/notifications.service.ts (hunk 5 above)
Comment (john-tco) [reply src/notifications/v2/notifications.service.ts:36]: ty :-)

**Review by dominicwest — CHANGES_REQUESTED** [review]
Also noticed we don't await the response for some sending of the emails, I know this isnt due to your PR, but this has caused performance issues for another job before. Lets prevent this now and add an await before any invocation of sendEmail

**Review by dominicwest — CHANGES_REQUESTED** [review]
Saved search scheduled jobs fail locally for me, I believe this is because of the cron jobs ive set them to run at and that this was a bug before. But ideally it shouldn't be so unstable. Falls over because it tries to insert a duplicate saved_search_notification before the other job clears it or something.

**Review by dominicwest — APPROVED** [review]
Class effort <3

**Pull request merged**